File: swift/medium/2-add-two-numbers.swift
// recursion
func addTwoNumbers(_ a: ListNode?, _ b: ListNode?) -> ListNode? {
    return addTwoNumbers(a, b, 0)
}

func addTwoNumbers(_ a: ListNode?, _ b: ListNode?, _ carry: Int) -> ListNode? {
    if (a == nil && b == nil) {
        return carry > 0 ? ListNode(carry) : nil
    }

    let vA = a?.val ?? 0
    let vB = b?.val ?? 0
    let sum = vA + vB + carry

    var node = ListNode(sum % 10)
    node.next = addTwoNumbers(a?.next, b?.next, sum / 10)
    return node
}

// iteration
func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    var a = l1
    var b = l2

    var dummy = ListNode(0)
    var current = dummy

    var carry = 0
    while (a != nil || b != nil) {
        let vA = a?.val ?? 0
        let vB = b?.val ?? 0
        let sum = vA + vB + carry

        carry = sum / 10
        current.next = ListNode(sum % 10)

        current = current.next!
        a = a?.next
        b = b?.next
    }

    if carry > 0 {
        current.next = ListNode(carry)
    }

    return dummy.next
}

File: swift/medium/2.add-two-numbers.java
// recursion
public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    return addTwoNumbers(l1, l2, 0);
}

public ListNode addTwoNumbers(ListNode l1, ListNode l2, int carry) {
    if (l1 == null && l2 == null) {
        return carry > 0 ? new ListNode(carry) : null;
    }

    int a = l1 != null ? l1.val : 0;
    int b = l2 != null ? l2.val : 0;
    int sum = a + b + carry;

    ListNode node = new ListNode(sum % 10);

    l1 = l1 != null ? l1.next : null;
    l2 = l2 != null ? l2.next : null;
    node.next = addTwoNumbers(l1, l2, sum / 10);
    return node;
}

// iteration
public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    int carry = 0;
    ListNode dummy = new ListNode(0);
    ListNode current = dummy;

    while (l1 != null || l2 != null) {
        int a = l1 != null ? l1.val : 0;
        int b = l2 != null ? l2.val : 0;

        int sum = a + b + carry;
        current.next = new ListNode(sum % 10);
        carry = sum / 10;

        l1 = l1 != null ? l1.next : null;
        l2 = l2 != null ? l2.next : null;
        current = current.next;
    }

    if (carry > 0) {
        current.next = new ListNode(carry);
    }

    return dummy.next;
}
